[2주차] 윤재홍 숫자 야구 게임 MVC 패턴 적용 제출합니다.

yoonjaehong26 opened the pull request:
안녕하십니까 그리디 3기 FE 윤재홍입니다. 
먼저 바쁘신 와중에 시간을 내주셔서 제 코드를 읽어주시고 리뷰해 주시는 김의천 리뷰어님께 진심으로 감사드립니다!
js를 배운 지 얼마 되지 않아 부족하지만 최대한 고민을 하고 코드를 짜 나가고 있습니다.
***
1주차에 이은 연계과제인 점을 고려하여 먼저 1주차에서 제가 최우선으로 고려했던 부분을 먼저 말씀드리겠습니다.

1.가독성과 유지보수를 위하여 함수단위로 코드를 구성
=>최대한 함수는 하나의 일만 할 수 있도록 고려하였고 모든 기능은 함수단위로 처리하도록 노력했습니다.

2.함수단위 코드를 통한 항후 유지보수와 확장성 고려
=>숫자야구게임을 어떻게 확장시킬까 고민하여 회원가입 비밀번호 설정 제한 알고리즘과 비슷하다고 느껴
 1주차 기준 파일 분활을 통하여 기존코드에서 기능 추가를 용이하게 해 보았고 클라스 생성자를 통하여 3자리수가 아닌 다른 자리수로 확장을 고려하였습니다.

3.에어비앤비 코드 컨벤션 준수를 통하여 보편적인 코드 작성
=>전체적인 코드 구조가 쉽게 이해가 잘 되게 또한 함수를 위주로 코드를 구상하였습니다.
***
1주차에 이어 2주차에서 중점적으로 고려한 사항들 말씀드리겠습니다.

1. 첫주차에 구상한 함수를 통한 유지보수성을 mcp패턴으로 확장하여도 유지가 가능한가
=>대부분의 시간을 이해하기 쉬운 mcv를 구상하려고 고민해 보았고 전체적인 코드구조의 변화를 한 커밋안에 담아 낼 수 있을 지 생각해 보아,
1주차의 함수위주의 코드를 통하여 model, controller,view 각각의 클라스에 함수의 위치를 재배치 함으로써 문제를 해결 해 보았습니다.

2.클라스 위주 코드에 대한 적응  
=>아직 클라스와 그 활용에 익숙하지 않아 문서들을 읽어보면서 익숙해졌고 코드에 적용해 보려 노력했습니다.
클라스에 대하여 더 깊은 내용들이 많아 더 고민해보고 공부해 볼 필요를 느끼고 있습니다.

3. ESlint 에어비앤비 컨벤션 준수를 통한 코드 보편화
=>mvc패턴 적용 이후 발생하는 eslint오류들을 검색해가며 왜 관련 규정이 생겼는 지 생각해 가며 코드를 리팩터링 해 나갔습니다.

***
김의천 리뷰어님께 질문드리고 싶은 점

1.mvc패턴에서 utils폴더 내  isValidInput(),numToArr() 두 함수의 적절한 위치에 대해 궁금합니다.
앞서 말씀드린 대로, isValidInput함수는 코드의 확장성을 위하여, numToArr함수는 코드 편의성을 위한 함수로,
BaseballGameModel class내부, model폴더 내부,현재와 같은 utils폴더 내부 세가지 방향성을 고민하였는데
mvc패턴에 어떤 방식이 적절한 코드구조인지 여쭈어보고 싶습니다!

2.아직 클라스에 익숙하지 않아 제가 적절하게 클라스를 사용하고 있는 지 궁금합니다.
현재 BaseballGameView클라스는 클라스에서 변하는 내용이 없어 인스턴스가 불필요하다 생각하여 클라스 그 자체로,
BaseballGameModel클라스는 인스턴스에 따라 확장성(digitCount와 같은 프로퍼티 변화)에 따라 인스턴스로 이용하고 있는데
제가 이해한 내용이 맞는 지, 가독성과 유지보수차원에서 적절한 지 여쭈어 보고 싶습니다.
***
다음은 기능구현 gif와 dev run test 결과입니다.
부족한 점이 많은 저의 PR을 읽어주셔서 감사드립니다.  
![화면 기록 2025-09-19 오후 7 36 17](https://github.com/user-attachments/assets/584f9968-5e3f-43f7-a8c9-3d799eddb45e)
<img width="941" height="607" alt="스크린샷 2025-09-19 오후 7 34 33" src="https://github.com/user-attachments/assets/5bd88ad1-6e1b-4326-b52a-035be347c936" />

Commit:
docs(README) : 기능 목록 추가

Commit:
feat(index.js) : 유효한 랜덤변수 생성

Commit:
feat(index.html,index.js) : 입력데이터 리스너생성

Commit:
feat(index.js) : 입력데이터 저장

Commit:
feat(checkInput.js):배열길이검사

Commit:
feat(checkInput.js):배열중복검사

Commit:
feat(checkInput.js):배열값타당성검사

Commit:
fix(index.js,checkInput.js):import오류수정

import대상범위 오류로 인해 import대상변경

Commit:
feat(index.js):입력데이터 유효성검사

Commit:
feat(index.js):야구게임 로직구현

Commit:
feat(index.js, uiChanger.js):ui변경

Commit:
feat(index.js):함수통합 밎 class구현

Commit:
fix(iddex.js,index.html):파일import오류수정

Commit:
feat(index.js):재시작기능구현

재시작기능이 없어 재시작시 변수재설정기능을 추가

Commit:
feat(index.js,uiChanger.js):재시작 ui초기화

Commit:
refacter(index.js,checkInput.js):데이터 자리수 소프트코드

코드 확장성을 위하여 숫자야구데이터의 자리수를 3에서 변수로 변경

Commit:
refacter(index.js,uiChanger.js):불명확한 함수명 변경

가독성과 통일성이 떨어지는 함수이름 변경:changeHrml=>changeUI,initEventListeners=>uiEventListeners,handlesummit=>gameLogic

Commit:
refactor(index.js):함수명 통일성 준수

 함수명의 통일일성을 위해 함수명 변경함 gmaeLogic-유지,  handleRestart=>restartLogic

Commit:
fix(index.js):restartLogic 함수명 정정

이전 커밋에서 바꾸지 못한 함수명 정정 uiEventListeners내
handleRestart=>restartLogic

Commit:
feat(index.js)함수문법 가독성개선

함수.bind(this)문법 가독성 감소하여 화살표함수로 대체

Commit:
refactor(index.js):변수 가독성을 위한 변수명 변경

변수 타입과 명확성의 부족으로 다음과 같이 변경
ball=>ballcount
strike=>strikecount

전체적은 변수명 통일을 위하여 다음과 같이 변수명 통일
result=>resultstring

Commit:
feat(index.js) 3스트라이크 시 메세지를 수정하여 게임완료를 명확하게 알림

Commit:
refactor(index.js): airbnb 코드 컨벤션에 따른 연산자 표현 변경

Commit:
refactor(index.js,uiChanger.js):함수목적성을 강조하기 위한 함수명 변경

Commit:
refactor(checkInput.js):불필요한 numToArr함수 import삭제

Commit:
refactor(index.js):에어비엔비 코드 컨벤션에 따른 연산자 좌우 빈칸 확보

Commit:
feat(checkInput.js):입력 유효성 검사 정규표현식 활용 조건 추가

Commit:
refactor(checkInput.js):중복된 숫자검증 제거

Commit:
fix(uiChanger.js):이전 ui 변경관련 커밋에서 반영 안된 부분 수정

이전 5a2de4fc7aa2800a6cbb6a9c4ab63e7dbbfd02cb커밋에서
 index.js getPlayResultString함수에서 바꾼 결과를 changeGameResultUI함수에서 반영이 안된 부분 수정

Commit:
feat(package-lock.json,package.json,.eslintrc.json):eslint기능추가

Commit:
refactor(index.js):에어비앤비 컨벤션에 따른 문자열 ''사용

Commit:
refactor(/src,index.html):MVC패턴 적용

model,view,controller폴더 생성 후 각 역할에 맞는 함수 위치 변경,할당
view,model의 경우 클라스 문법을 사용하여 코드의 모듈화에 집중함
전체적인 코드의 구조변화 리펙터링임에 주의하여 코드구조변경 외의 다른 기능추가나
다른 영역 리펙터링을 하지 않음

Commit:
refactor(additionalUtilFunction.js,BaseballGameView.js):에어비앤비 컨벤션에 따른 작은 문자열 따움표 사용

Commit:
refactor(BaseballGameModel.js,additionalUtilFunction.js,checkInput.js):에어비엔비 컨벤션에 따른 함수,클라스 마지막 라인 확보

Commit:
refactor(baseballGameIndex.js,BaseballGameModel.js):에어비엔비 컨벤션에 따른 depth통일화,화살표함수 매개변수 괄호사용

Commit:
feat(.eslintrc.json):eslintrc 규칙추가

1.

Commit:
refactor(checkInput.js):에어비앤비 컨벤션에 따른 for ~of,isNaN문법 대체

Commit:
refactor(baseballGameIndex.js,BaseballGameView.js):BaseballGameView클라스 호출 구조 변경

BaseballGameView클라스는 인스턴스에 따라 달라지지 않는 자체적인 클라스이므로 static 매서드 사용,호출방식 변경

Commit:
refactor(BaseballGameModel.js):연산자 문법 수정,하드코딩 된 매직넘버 변수할당

에어비앤비 컨벤션에 따른 연산자 좌우 빈칸 확보,유지보수를 위한 하드코딩 제거

Commit:
refactor(baseballGameIndex):BaseballGame클라스 생성자 기능분리밎 인스턴스 기능 업데이트

yoonjaehong26 committed:
Merge branch 'yoonjaehong26' into main

Commit:
fix(checkInput,index.js,uiChanger.js,util.js):이전 merge에서 잘못 추가 된 코드 삭제

Commit:
refactor(baseballGameIndex.js, BaseballGameView.js):view에서 게임완료 조건처리를 controller로 이전

wzrabbit commented on src/model/BaseballGameModel.js:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    if (strikeCount === this.digitCount) {
+      resultString = '🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?';
```
`'🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?'` 메시지의 경우 Model이 가지고 있기에는 꽤 구체적인 정보 같다는 생각이 들었어요. 이 정보까지 Model이 알기에는 Model이 하는 역할이 큰 것 같고, 이후 출력할 문장을 바꿀 때는 출력과 관련된 로직만을 변경함에도 불구하고 Model을 수정해야 하는 상황이 올 것 같아요.

- `BaseballGameModel.play`는 각 라운드의 숫자야구 결과(스트라이크, 볼 개수)를 구하는 본래 역할에 충실하고
- `BaseballGameModel.play`이 반환한 결과로 구체적으로 무얼 할 지는 다른 도메인이 결정하고 담당하는 것

은 어떨까요?

yoonjaehong26 replied:
넵 1주차때에 html에 직접 dom을 조작할까 했지만 코드가 복잡해져서 문자열로 ux를 조정했습니다.
지금 와서 생각해 보니 mvc패턴을 적용하여 controller에서 객체로 정보를 넘겨주어 view만 조작하면 되어서
처음부터 mvc패턴을 적용했을 시에 유지보수에 더 유리했을 것 같다고 느낍니다!
***
```
refactor(baseballGameIndex.js,BaseballGameModel.js):play()기능분리 밎 확장된 calculatePlayResult 함수 추가

play()함수를 calculatePlayResult()대체하여 객체를 통한 보편적인 반환을 가진 model함수를 고안하였으나
과제 요건을 고려하여 play함수를 문자열 반환하는 추가기능을 가진 함수로 사용.
```

```
refactor(BaseballGameView.js,BaseballGameModel.js,BaseballGameView.js):게임완료 문자열 처리 구역 변경

게임완료 문자열 🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?
이 model에 있는 것이 mvc패턴에 적합하지 않는다 판단하여 control에서 처리하고
같은 이유로 view에서도 같은 문자열 control에서 처리함
```
위 두 커밋을 통하여 modell은 객체를 반환하여 유동적인 함수를 제작하였고, control에서 객체를 처리하여 최종적으로 view에서
문자열을 처리하도록 바꾸었습니다.

mvc패턴 적용을 통해 코드 구상을 고민해 볼 수 있었던 의미깊은 과정이었습니다 감사합니다!

wzrabbit replied:
오, 이제 Model은 자신의 관심사 내에서의 값만 반환하고, 받은 값을 Controller가 판단하여 적절한 View를 호출하는군요 👍🏻

wzrabbit commented on src/view/BaseballGameView.js:
```diff
@@ -0,0 +1,25 @@
+
+  static alertWrongUserInput() {
+    alert('잘못된 입력입니다. 중복되지 않는 서로 다른 3개의 숫자를 입력하세요');
+    document.getElementById('user-input').value = '';
```
👍🏻 이런 처리 좋네요

wzrabbit commented on src/controller/baseballGameIndex.js:
```diff
@@ -0,0 +1,51 @@
+    BaseballGameView.resetGameResultUI();
+  }
+
+  uiEventListeners() {
+    document.getElementById('submit').addEventListener('click', this.gameLogic);
+    document.getElementById('game-restart-button').addEventListener('click', this.restartLogic);
+  }
+
+  gameLogic = (e) => {
+    e.preventDefault();
+
+    const userInput = document.getElementById('user-input').value;
+    if (!isValidInput(userInput, this.digitCount)) {
+      BaseballGameView.alertWrongUserInput();
+      return;
+    }
+
+    const userInputNumbers = numToArr(userInput);
+    const resultString = this.BaseballGameModel.play(this.computerInputNumbers, userInputNumbers);
+
+    if (resultString === '🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?') {
+      BaseballGameView.changeCompleteGameResultUI(resultString);
+    } else {
+      BaseballGameView.changeIncompleteGameResultUI(resultString);
+    }
+  };
+
+  restartLogic = () => {
+    this.computerInputNumbers = this.BaseballGameModel.makeRandomNumber();
+    BaseballGameView.resetGameResultUI();
+  };
```
메서드명들을 모두 명사로 두신 이유가 있으신가요?

yoonjaehong26 replied:
다른 함수들은 모두 동사+명사의 구조로 함수명을 지었으나,
`uiEventListeners()`의 경우 함수라기 보단 dom리스너의 모음이라 함수의 느낌이 나지 않았고
`gameLogic()`,`restartLogic()`의 경우 또한 아직 화살표함수가 함수라는 
느낌이 들지 않아 깊은 고민 없이 그대로 두었던 것 같습니다..
***
이번에 클라스에 대하여 공부하면서 ,매서드의 이름을 통해 클라스 내에서 
전체적인 흐름을 파악하는 것의 중요성을 느끼게 되어 말씀해주신 함수 작명법과 통일성을 고려하여 다음과 같이 변경하였습니다!
***
```
uiEventListeners  => ListenUIEvent
gameLogic => handleGuessSubmit
restartLogic => restartGame
```

wzrabbit replied:
화살표 함수를 처음에 접하시는 입장에서는 꽤 생소할 수 있을 거라 생각합니다. 저도 처음 접할 때는 신기하게 느껴지는 문법이면서도 낯설었던 것 같아요. 하지만 형태가 다르더라도, 화살표 함수 내부에서 다뤄지는 내용은 여전히 "동작", "행동" 과 밀접한 관련이 있어요. (물론 몇몇 상황에서는 명사를 사용할 수도 있지만) 웬만한 상황에서는 동사의 형태일 때 

> uiEventListeners()의 경우 함수라기 보단 dom리스너의 모음이라 함수의 느낌이 나지 않았고

이 경우에는 음, 이렇게 생각하시면 어떨까요?
- 우선 `uiEventListeners` 함수 내부의 `addEventListener`는 리스너를 등록시키는 **동작**이에요. `addEventListener`의 이름을 보시면 아실 수도 있는데, 이름이 `eventListener`가 아니라 `addEventListener`에요.
- `addEventListener`가 여러 개 담긴 `uiEventListeners` 역시, 내부의 두 리스너를 등록시키는 **동작**을 수행하기 때문에 함수가 하는 역할에 매우 가깝다고 할 수 있어요.

단순히 무언가의 모음이라고 생각되어 함수의 느낌이 잘 나지 않았다면, 함수의 느낌이 아닌 예도 들어볼게요. 사실 이전 리뷰에서 예를 들었던 계산기 오브젝트를 다시 가져왔어요.

이번에는 이 오브젝트의 이름은 동사형이 아니죠? 아마 이러한 형태가 재홍님께서 생각하시는 "모음" 에 더 가까운 형태일 거에요. 여기에서는 `calculator` 자체가 무언가를 실행시키는 행동이 아니고, 여러 함수들, 그러니까 기능들을 담은 객체의 형태거든요. 이외에도 지금까지 구현하신 클래스들도 해당해요.

```js
const calculator = {
  add(number1, number2) {
    return number1 + number2;
  },
  subtract(number1, number2) {
    return number1 - number2;
  }
};
```

**여러 함수들을 실행해 더 큰 단위의 작업을 하는 것**과, **함수들을 모아두는 객체**의 차이라고 생각하시면 될 것 같아요.

wzrabbit replied:
그리고, 새로운 이름들 정말 마음에 들어요! 함수의 경우 지금처럼 **동작** 을 잘 설명하는 이름이면 좋을 것이라 생각해요. 이름에 대한 고민은 앞으로도 계속될 것이랍니다

wzrabbit commented on src/model/BaseballGameModel.js:
```diff
@@ -0,0 +1,51 @@
+    this.digitCount = digitCount;
+  }
+
+  makeRandomNumber() {
+    const numbers = [];
+    while (numbers.length < this.digitCount) {
+      const number = MissionUtils.Random.pickNumberInRange(1, 9);
+      if (!numbers.includes(number)) {
+        numbers.push(number);
+      }
+    }
+    return numbers;
+  }
```
메서드명이 `makeRandomNumber`인지라 내부 구조를 보지 않았을 때는 `213`과 같은 값을 예상했는데, 그런데 실제로는 `[2, 1, 3]`과 같은 값을 반환하는 점에서 메서드명과 실제로 하는 일이 다른듯한 느낌이 들었어요. 또한 메서드명은 단수형인데 반환하는 값의 변수명은 `numbers`로 복수형인 것도 있고요?

사용하는 입장에서 메서드명만 보고 역할을 예측하기 쉽도록 네이밍 개선을 요청드리고 싶어요!

yoonjaehong26 replied:
변수명을 통하여 리턴값을 예상 할 수 있게 구성하는 것은 
함수를 더 잘 이해할 수 있게 큰 도움이 될 것 같아요!
***
 `makeRandomNumber()` 에서`makeRandomNumbers()`로 변경하였습니다.
다른 리턴값이 있는 함수들도 함수명 확인 완료했습니다. 감사합니다!

wzrabbit replied:
이제 함수명만 보고 값이 여러 개 반환된다고 당황하는 일은 없을 것 같군요! 💪🏻 지금 이름 정도면 충분히 좋다고 생각해요.

여기에 더해, `makeRandomNumbers`가 반환하는 수는 항상 한 자리 수인 것을 이용해, `makeRandomDigits` 와 같이 **의미를 좁힐 수도** 있을 거에요.

사실 이전에는 `makeBaseballGameRandomNumbers` 와 같이 "숫자야구 게임" 의 수임이 드러나는 이름도 추천했었는데, 이번 상황에서는 그렇지 않아도 될 것 같다는 생각이 들었어요. 왜냐하면, 재홍님께서 클래스의 이름을 `BaseballGameModel` 이라고 이미 지어두셨기 때문이에요. **클래스의 이름 자체에서 의미가 충분히 설명되어 메서드명을 좀 더 간략하게 축약할 수 있는 상황**이 있어요. 앞으로 이름을 지으실 때 참고하시면 좋을 것 같습니다!

wzrabbit commented on src/view/BaseballGameView.js:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  static alertWrongUserInput() {
+    alert('잘못된 입력입니다. 중복되지 않는 서로 다른 3개의 숫자를 입력하세요');
```
작성하신 숫자 야구 게임은 다른 자릿수로도 플레이할 수 있도록 확장 가능하도록 열어두신 것 같아요. 그렇다면 이 부분의 메시지도 확장 가능하게 열어보시는 것은 어떨까요?

yoonjaehong26 replied:
`fix(baseballGameIndex.js,BaseballGameView.js):숫자야구 자리수 변동에 대한 미구현 분제 해결 `
위 커밋에서 자리수 변동에 대한 실제 작동 가능하게 수정하였습니다!
***
'1주차에서 이정도면 확장이 가능하겠지' 라는 생각으로 구상하긴 하였는데
말씀해 주신 대로 부족한 부분이 많았습니다.

하지만 덕분에 확장성을 고려할 때 실제로 다른 기능을 생각해 보고 코드를 짜 보면서
사용성이 좋은 코드를 생각할 수 있게 되었습니다.
좋은 방법 알려주셔서  감사드립니다! 

```js
  static alertWrongUserInput(digitCount) {
    alert(`잘못된 입력입니다. 중복되지 않는 서로 다른 ${digitCount}개의 숫자를 입력하세요`);
```

wzrabbit commented on src/controller/baseballGameIndex.js:
```diff
@@ -0,0 +1,51 @@
+import isValidInput from '../utils/checkInput.js';
+import numToArr from '../utils/additionalUtilFunction.js';
+
+export default class BaseballGame {
```
보고 궁금증이 많이 생긴 클래스군요 🤔 질문에도 적어주셨듯이 이 클래스에 대해 고민하고 계심을 알아요.

## 1️⃣ 
`BaseballGame`의 경우 `constructor`가 있으면서 동시에 인스턴스를 만들지 않고도 사용할 수 있는 `start` 메서드가 있습니다. 이 클래스를 실행시키기 위해서는 인스턴스를 먼저 만든다음 그 인스턴스를 사용해야 하는 건가요, 아니면 인스턴스를 만들지 않고 바로 `start` 메서드를 실행시켜야 하는 것일까요? 사용법이 다소 헷갈리게 느껴졌습니다. 

클래스를 사용하는 몇몇 코드들을 한 번 훑어보시는 것도 도움이 될 것 같아요, 클래스를 처음 배우시는 입장이라고 말씀하셨는데, 이 경우에는 예시를 좀 많이 보시면 될 것 같아요.

## 2️⃣ 
`constuctor`의 `digitCount` 파라미터를 통해 이 클래스의 확장을 고려해주신 것으로 이해했습니다. 그런데 맨 아랫줄의 사용례를 보니 실제로는 `BaseballGame.start()`을 통해 컨트롤러를 실행시키시는 것으로 보이며, `start` 메서드 내에서는 항상 `BaseballGame`에 대한 인스턴스를 파라미터 없이 생성하기에 항상 3자리 수로만 숫자야구 게임을 할 수 있어보이는데요 🤔 3자리 수가 아닌 다른 자릿수의 숫자야구 게임으로 바꾸도록 세팅을 바꾸려면 어떻게 바꿔주어야 하나요?

## 3️⃣ 
`start` 메서드를 static으로 선언해 주신 이유가 궁금합니다!

yoonjaehong26 replied:
지금 다시 보니 저조차도 이해하기 힘든 구조인 것 같습니다.
분명 pr제출 시에는 괜찮다 생각하였는데 저라도 처음 이 코드를 보면 당황스러울 것 같습니다...
***
우선 이러한 코드를 짜게 된 경위는, eslint를 설치한 지 얼마 안된 상태에서 클라스를 함께 습득하다 보니 자꾸 this 아니면 static을 쓰라는 경고가 떠서
매서드가 인스턴스에서 쓰이면 this, 클라스 자체로 쓰이면 stastic이라는 이지선다에 
매몰되어 전자는 생성자에, 후자는 start으로 배치했던 것 같습니다.

말씀하신 대로 클라스를 더 공부하고 나니 그냥 생성자에 통일하는 것이 깔끔하다고 
생각합니다..
***
위의 이유로 인스턴스에 상관이 없는 start()에 static으로 선언하였고이상한 구조가 되었습니다.
***
이후 아래와 같은 깔끔한 구조로 바꾸었으며 자리수 변경에도 용이하게 변경하였습니다.

```js
constructor(digitCount = BaseballGame.DEFAULT_DIGIT_COUNT) {
    this.digitCount = digitCount;

    this.BaseballGameModel = new BaseballGameModel(this.digitCount);

    this.isCompleteGame = false;
    this.computerInputNumbers = this.BaseballGameModel.makeRandomNumbers();
    this.listenUIEvent();
    BaseballGameView.resetGameResultUI();
  }
```

```js
const game = new BaseballGame(3);
```

wzrabbit replied:
아마 ESLint가 하는 경고로 인해서 많이 혼란스러우셨을 것 같다는 생각이 들었어요. ESLint는 코드에 있을만한 잠재적인 위험이나 권장되지 않는 코드를 작성하는 경우 이렇게 에러를 띄워주는 경우가 많아요. 따라서 준수하시도록 노력하셔야 해요.

하지만 ESLint의 권장사항을 따라가다 보면 가끔 코드가 더 알아보기 힘든 형태로 바뀌기도 할 거에요. ESLint는 권장되지 않는 코드 A를 B와 같이 바꿀 것을 제안했는데, 경고하는 의미가 잘 파악되지 않는 입장에서는 의도치 않은 C로 바꾸는 상황이 생기게 돼요.

ESLint의 권장사항을 따라가다가 막히셨다면, 바꾸기 전보다 코드의 의미와 의도를 더 알아보기 힘들다고 생각하신다면, 기존에 에러가 났었던 의도하신 코드로 구현하시거나 이 코드를 PR을 적으신 후, 리뷰어에게 도움을 요청해보는 것도 좋은 방법일 수 있을 것 같아요. 특히 PR에 헤맸던 이야기를 적어주시면 리뷰어가 더 상황을 이해하기도 좋을 테고요!

wzrabbit replied:
전반적으로 `static`을 많이 사용해주셨는데, `static`의 쓰임새는 주로 "인스턴스 생성 없이 클래스에서 특정 기능을 가져다가 사용할 수 있도록 고안하고 싶은 경우" 에 좋아요.

반면 클래스를 통해 생성한 인스턴스들이 내부적으로 지니는 값에 따라 연산을 수행하거나, 각 인스턴스마다 해당 함수를 가지도록 구현하고 싶으신 경우에는 `static`이 적합하지 않을 가능성이 높아요. 이 경우 일반 메서드를 사용해요. `static`은 클래스가 인스턴스와 관계없이 공용으로 지니는 메서드라고 생각하시면 될 거에요.

클래스라면 흔하게 예를 드는 "자동차" 클래스로 이야기해볼게요.
- 자동차는 각각 이름을 가지고 있어요. 자동차의 이름을 불러오는 함수가 실행되면 자동차 이름을 반환해요.
- 자동차 클래스 자체에는 자동차의 구매연도와 수명을 인자로 받으면 자동차의 폐기 예상 연도를 반환하는 함수가 있어요.

첫 번째 함수는 인스턴스마다 지니는 자동차 이름이 다르므로 일반 메서드로 작성하는 것이 적합하고, 두 번째 함수는 인스턴스랑 상관없이 자동차의 구매연도와 수명을 받아 계산 결과를 반환하는 일만을 하므로 `static` 메서드로 작성하는 것이 주로 적합할 거에요.

이를 토대로, 숫자야구 게임 관련 클래스에서 각 함수에 `static`을 적용할지, 어떤 값을 상수로 저장하고 어떤 값을 생성자로써 받을지를 고민해보실 수 있을 거에요. 생성자로 값을 받는 형식으로 인스턴스를 만들면 인스턴스마다 서로 다른 값을 지니니 같은 클래스에서 나온 인스턴스더라도 필요에 따라 서로 다른 역할을 하도록 활용하실 수 있을 거에요. 반면 특정 필드를 상수로 고정해두었다면, 상수는 불변값이니 모든 인스턴스가 필요에 따라 그 통일된 상수를 적절히 활용할 수도 있을 테고요.

wzrabbit commented on src/model/BaseballGameModel.js:
## 👍🏻 
전체적으로 사용법이 명확하고 각 메서드가 각자의 할 일만을 충실하게 이행하도록 잘 구현해주신 것 같아요! 인스턴스를 생성할 시 숫자야구 게임의 자릿수를 통해 게임의 모드를 결정하고, 그 이후 `play`를 통해 본격적인 기능을 사용한다는 점, 그리고 그 외에 쓰인 메서드들은 각각 자신의 기능만 수행해 복잡한 기능을 수행하는 `play`를 서포트하는 구조가 마음에 드네요

## ➕ 
지금도 좋지만, 여기 **캡슐화**가 더해지면 어떨까 생각해요. 캡슐화란 외부로부터 내부를 감싸 은닉시키는 것을 의미해요.

재홍님께서 상상하신 이 모델 클래스의 사용법은 바로 인스턴스를 만들고, `play` 메서드만 실행시켜 결과를 반환받는 것이라고 예상해요. 그런데 외부에서 함부로 `makeRandomNumber`를 게임 중 사용하는 등의 상황이 생기거나, `digitCount`의 값을 `"우하하"`와 같은 잘못된 값으로 변경하는 등, **누군가 의도 밖의 사용 방법대로 클래스를 다룬다면, 기능이 쉽게 오염되어 오작동할 수도 있을 것 같아요.**

사용처에서는 `play`와 같은 메서드와 같이 사용이 의도된 메서드들만 드러내고, 내부적으로만 사용되는 메서드 및 필드들은 철저히 숨겨두어, **잘못된 사용으로 인한 버그를 사전에 방지하고, 사용법을 명확히** 하는 것은 어떻게 생각하시나요? 사용처의 지시가 필요 없는 부분의 기능은 **메서드끼리 메시지를 주고받도록** 하는 거에요. JavaScript에서도 이를 구현할 방법은 있을 거에요.

비유하자면, 손님이 직접 주방에서 요리하고 있는 요리사들에게 일일이 지시해 잘못된 요리가 나오게 하기보다는, 손님은 종업원에게만 지시하고 주방에서 일어나는 일은 종업원이 각각의 요리사들에게 지시할 수 있도록 해 보는 것을 이야기한 거에요. 어떻게 생각하시나요?

yoonjaehong26 replied:
넵 캡슐화에 대하여 알아보고 제가 생각한 캡슐화는, '필요한 부분만 내보내고 나머지는 전부 숨긴다'로 이해했고 '작은 프로그래밍'으로 이해한 클라스를 더 체계적으로 만들어주는 과정으로 이해했습니다! 따라서 필요한 프로퍼티만 호출할 수 있게 하여 
`refactor(baseballGameIndex.js, BaseballGameModel.js, BaseballGameView.js):클라스내 변수,매서드 프로퍼티 캡슐화`
위 커밋 과정에서 캡슐화를 진행하여 은닉화를 하였고 검색하며 클로저 개념과 가비지 컬랙터에 대해 잘 이해할 수 있었습니다.

wzrabbit commented on src/view/BaseballGameView.js:
오 전체적으로 이 모듈에서의 View는 출력에 필요한 값만을 받아와 오로지 출력만 담당하는군요, 군더더기 없이 출력에만 집중하는 로직같아 깔끔해보입니다!

클래스 + static을 사용하는 현재의 예시도 좋고, 지금 View는 가지고 있는 상태가 하나도 없잖아요? 파라미터로 입력을 받으면 그 값을 처리하고, 그러고 잊잖아요? 이 경우면 오브젝트를 이용하는 방법도 하나의 방법이 될 수 있을 것 같아요.

아래는 간단한 사칙연산을 하는 Model의 예시에요. 역시 상태를 지니지 않으면서, 여러 개의 관심사가 비슷한 함수를 오브젝트를 통해 묶은 형태에요. 저는 두 형태 모두 좋다고 생각하니 상황에 따라 원하시는 방향으로 적용해보세요

```js
const calculator = {
  add(number1, number2) {
    return number1 + number2;
  },
  subtract(number1, number2) {
    return number1 - number2;
  }
};
```

yoonjaehong26 replied:
```js
const BaseballGameView = {

  changeIncompleteGameResultUI(resultString) {
    document.getElementById('result').style.display = 'block';
    document.getElementById('result').innerHTML = resultString;
    document.getElementById('game-restart-button').style.display = 'none';
  },

  changeCompleteGameResultUI() {
    document.getElementById('result').style.display = 'block';
    document.getElementById('result').innerHTML = '🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?';
    document.getElementById('game-restart-button').style.display = 'block';
  },

  resetGameResultUI() {
    document.getElementById('user-input').value = '';
    document.getElementById('result').innerHTML = '';
    document.getElementById('result').style.display = 'none';
    document.getElementById('game-restart-button').style.display = 'none';
  },

  alertWrongUserInput(digitCount) {
    alert(`잘못된 입력입니다. 중복되지 않는 서로 다른 ${digitCount}개의 숫자를 입력하세요`);
    document.getElementById('user-input').value = '';
  },
};
export default BaseballGameView;
```
네 위와 같이 한번 코드를 짜 봤습니다. 상태를 가지지 않으면 복잡하지 않은 구조로 깔끔하게 코드를 구성할 수 있는 것 같습니다!
***
저는 dom탐색을 줄이고 결과 문자열을 넣어 보고 싶어서 클라스로 구성해 보았습니다!
다양한 구조를 알려주셔서 감사합니다!
```js
export default class BaseballGameView {
  static #GAME_RESULT_STRING = '🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?';

  static #result = document.getElementById('result');

  static #userInput = document.getElementById('user-input');

  static #restartButton = document.getElementById('game-restart-button');

  static changeIncompleteGameResultUI(resultString) {
    this.#result.style.display = 'block';
    this.#result.innerHTML = resultString;
    this.#restartButton.style.display = 'none';
  }

  static changeCompleteGameResultUI() {
    this.#result.style.display = 'block';
    this.#result.innerHTML = this.#GAME_RESULT_STRING;
    this.#restartButton.style.display = 'block';
  }

  static resetGameResultUI() {
    this.#userInput.value = '';
    this.#result.innerHTML = '';
    this.#result.style.display = 'none';
    this.#restartButton.style.display = 'none';
  }

  static alertWrongUserInput(digitCount) {
    alert(`잘못된 입력입니다. 중복되지 않는 서로 다른 ${digitCount}개의 숫자를 입력하세요`);
    this.#userInput.value = '';
  }
}
```

wzrabbit replied:
훨씬 나아진 것 같아요, 좋습니다 👍🏻 
작성하신 기능들의 경우 모두 static 키워드를 사용해서 만들 수 있겠네요

wzrabbit commented on index.html:
```diff
@@ -24,7 +24,7 @@ <h3>📄 결과</h3>
+  <script type="module" src="src/controller/baseballGameIndex.js"></script>
 </body>
 
 </html>
```
<img width="180" height="83" alt="image" src="https://github.com/user-attachments/assets/ad0f43fa-2f71-4cd8-8635-1baa2b210030" />

여기 있는 진입금지 표시같이 생긴 건 무엇일까요? 🤔

yoonjaehong26 replied:
검색하고 알아보니,  github에서 지원하는 ui로
코드가 수정되거나 제거된 부분에 이전 댓글이 달렸다면 혼선을 방지하고자 
이전 댓글을 보이지 않게 처리하고 나타나는 ui라고 합니다.
***
pr도중 커밋으로 해당 코드를 변경하여 해결하였다는 이정표로 생긴 기능인 것 같습니다.

wzrabbit replied:
아쉽지만 말씀하신 특징과 실제로 위의 "진입금지 표시" 의 의미는 서로 연관성은 없어요. 이 표시는 **파일의 마지막에 개행이 포함되지 않음**을 경고하는 표시에요. 영어로는 "no newline at a end of file" 이니, 영어 키워드로 구글링을 해 보시면 더 자세한 정보들을 찾으실 수 있을 거에요.

대부분의 경우 프로그램은 올바르게 작동하겠지만, 혹시 모를 잠재적인 문제를 방지하기 위해 파일의 마지막은 개행으로 끝나도록 코드를 작성하실 것을 추천드려요.

보다 자세한 정보를 알려드리기 위해 아래의 글을 추천드릴게요.

- https://hyeon9mak.github.io/github-no-newline-at-a-end-of-file/

Commit:
refactor(baseballGameIndex.js):햇갈릴 수 있는 숫자야구 자리수 상수처리

3이 맥략이 없으면 매직넘버로 보일 수 있어 상수로 변환

wzrabbit reviewed (changes requested):
@yoonjaehong26 👋🏻

안녕하세요, 재홍님! 2주차 미션을 함께하게 된 리뷰어 김의천이라고 합니다. 만나뵙게 되어 반가워요 😊
기능 추가 없이 이렇게 리팩터링만 해 보는 미션이라니 좀 신기하죠? MVC 패턴을 적용해 보시니까 어떤가요? 더 코드가 깔끔해진 것 같나요, 아니면 반대로 더 복잡해졌나요?

<img width="150" height="135" alt="image" src="https://github.com/user-attachments/assets/9f68a509-314b-49b8-aa22-c4f6678f0acc" />


## 1️⃣ 질문/답변

> 1.가독성과 유지보수를 위하여 함수단위로 코드를 구성
> =>최대한 함수는 하나의 일만 할 수 있도록 고려하였고 모든 기능은 함수단위로 처리하도록 노력했습니다.

특히 모델에서 하나의 복잡한 기능을 여러 개로 나누어서 관리하신 것 봤어요, 잘 하셨습니다. 앞으로도 하나의 기능이 방대해진다면 지금처럼 필요에 따라 나누는 연습을 계속 하시면 좋을 것 같습니다.

> 1주차 기준 파일 분활을 통하여 기존코드에서 기능 추가를 용이하게 해 보았고 클라스 생성자를 통하여 3자리수가 아닌 다른 자리수로 확장을 고려하였습니다.

미래를 예상하고 확장을 해 보시는 연습은 정말 좋습니다. 다만 확장을 대비하셨다면 실제로 확장을 해 보았을 때 나머지 기능도 모두 올바르게 작동하는 지도 확인해 보시면 좋을 것 같아요. 숫자야구를 네 자리 수를 받도록 바꾼 다음, 모든 기능이 제대로 작동하는지 확인해 보시면 어떨까요? 기본 기능뿐만 아니라 여러 가지 상황에서 기능이 올바르게 작동하는지, 어색한 UI는 없는지 확인해 보시는 연습을 해 보시면 좋을 것 같습니다.

> 1.mvc패턴에서 utils폴더 내 isValidInput(),numToArr() 두 함수의 적절한 위치에 대해 궁금합니다.
> 앞서 말씀드린 대로, isValidInput함수는 코드의 확장성을 위하여, numToArr함수는 코드 편의성을 위한 함수로,
> BaseballGameModel class내부, model폴더 내부,현재와 같은 utils폴더 내부 세가지 방향성을 고민하였는데
> mvc패턴에 어떤 방식이 적절한 코드구조인지 여쭈어보고 싶습니다!

`utils` 폴더명을 보니 유틸 함수(utility function)이 떠오르는데요, 재홍님은 유틸 함수가 무엇이라고 생각하시나요?

저는 유틸 함수를 **특정 비즈니스 로직에 구애받지 않는 일반적이고 범용적인 기능을 지닌 함수**라 생각해요. 프로젝트 내의 여러 도메인/주제를 지니는 함수 어디서든 사용될 수 있고, 심지어는 여러 프로젝트에서도 바로 사용될 수 있는 함수들이요.

- ex) 날짜의 형식을 변경하는 함수.

유틸 함수는 Model, View, Controller 모두에서 범용적으로 유용하게 사용될 수 있으면서 주제를 해치지 않기에 `utils` 디렉토리와 같이 별도로 분리하고 자유로이 사용할 수 있다고 생각해요.

반면 분리할 수 있는 기능을 가지더라도 그 기능이 특정 프로젝트의 도메인에 종속된다면 유틸 함수보다는 **도메인 함수(domain function)** 라 생각해요. 이러한 함수들은 모델의 성격을 강하게 띄는 경우가 많아, 유틸 함수처럼 별도의 디렉토리로 분리하기보다는, Model을 보조하는 함수로써 쓰이는 경우가 많은 것 같아요. Model과 View가 서로를 아는 것이 안 될뿐, 큰 기능을 담당하는 Model이 다른 Model의 보조를 받는 것은 문제 없으며 오히려 자주 이렇게 응용돼요. 재홍님이 큰 기능을 지닌 하나의 함수를 여러 개의 함수로 분리하신 것처럼요.

Model의 보조를 받는 도메인 함수라면 같은 파일에 둘까 다른 파일에 둘까를 고민하실 수 있는데, 해당 함수가 여러 곳에서 재사용되는가, 특정 Model과 관심사가 강하게 결합되어 있는가 등을 보시면 어떨까 생각합니다. 여기서부터는 정답이 없다고 생각해요.

`isValidInput`, `numToArr`. 재홍님이 생각하시기에는 이 함수들이 하는 일은 어느 범위라고 생각하시나요? 특정 도메인에 종속되어 기능을 보조하나요? 아니면 당장 다른 프로젝트에서 쓰여도 무방한, 특정 주제를 지니지 않는 범용적인 일을 하나요?

> 2.아직 클라스에 익숙하지 않아 제가 적절하게 클라스를 사용하고 있는 지 궁금합니다.
> 현재 BaseballGameView클라스는 클라스에서 변하는 내용이 없어 인스턴스가 불필요하다 생각하여 클라스 그 자체로,
> BaseballGameModel클라스는 인스턴스에 따라 확장성(digitCount와 같은 프로퍼티 변화)에 따라 인스턴스로 이용하고 있는데
> 제가 이해한 내용이 맞는 지, 가독성과 유지보수차원에서 적절한 지 여쭈어 보고 싶습니다.

총 세 개의 클래스를 사용하셨는데 모두 코멘트로 제가 느낀 점과 의견 적어볼게요!

## 2️⃣ 리뷰

### 기능 면에서

1주차에서 한 단계 거친 것도 있어서 그런지, 세 자리 수에서 숫자야구 로직은 잘 작동하는 것 같아요. 이번에는 조금 더 고도화를 해 보는 것은 어떨까요?

- 재홍님의 숫자야구 코드는 세 자리 수 뿐만 아니라 다른 자릿수로 게임을 하는 확장성까지 고려하고 있어요. 다른 자릿수로 게임을 할 수 있도록 다른 프로그래머가 임의로 코드를 바꾸었을 때도 군더더기 없이 잘 작동하도록 점검을 추천드려요. 지금은 UI를 보았을 때 몇몇 부분에서 불일치가 보이는 것 같았어요.

- 또한 게임이 끝났을 때도 여전히 사용자가 수를 입력할 수 있었고, 틀린 수를 입력하면 다시 재시작 버튼이 없어지고 다른 결과가 보이는데 이 부분이 개인적으로 어색하다고 느껴졌어요. 재홍님도 저와 똑같은 생각을 하고 계신다면 이 부분을 개선해보시는 것은 어떨까 생각합니다.

- 1주차 첫 제출에는 3스트라이크와 함께 재시작 버튼만 보이도록 구현하셨음을 알아요. 지금은 명확하게 클리어했음이 보이는 결과창이 보이는데 잘 하셨어요. 이렇게 한 단계씩 저희 개선해 봅시다.

간단한 게임이지만 배워갈 점과 개선할 점은 저희가 고민하면 할수록 무궁무진하게 많다고 생각해요.

### 네이밍 면에서

함수명인데 변수명에 더 가까워 보이는 네이밍도 있었던 것 같았고, 이름을 보았을 때 실제로 하는 작업과 달라보이는 이름도 있어보여 개선해 볼만한 점들이 보였어요. 하지만 1주차와 비교했을 때 훨씬 개선된 점도 보였습니다. 네이밍에 대해서는 계속해서 저도 고민해 볼 테니, **기능을 명확하게 설명하는** 멋진 이름 저희 생각해 보아요!

"이름은 길어져도 좋으니 명확한 이름을 지어보자" 라는 말씀은 이미 들으셨을 것 같으니 저는 다음 단계에 해당하는 이야기를 해 보고 싶어요. 으음... 제가 이름을 지을 때 어떻게 짓는지를 조금 적어보아요. 절대적인 진리라고 보지 마시고, 한 개발자의 의견이자 노하우라고 생각해보시면 어떨까요?

1. 저는 먼저 함수가 하는 일을 직관적으로 한글로 풀어 적어요. 함수의 기능을 적기 위해 한글 주석으로 적었다면 그 주석이 딱 어울리는 상황이겠네요.

2. 이제 제가 풀어쓴 한글 설명의 모든 특징을 담는 영어 문장이나, 영어 단어의 조합으로 적어보아요. 필요시 영어 사전도 참고하고 있어요. 여기서는 길이를 생각하지 않고 **의미의 정확도를 1순위로 생각해요.** 이렇게 하면 어색하고 이름이 길 수는 있지만 함수가 하는 일 자체는 정확한 이름이 될 거라고 보아요.

3. 이제 여기서 **의미를 좁히지 않는 선에서 줄일 수 있는 단어를 줄여보거나 생략해** 보고 있어요. 여기서부터는 이름의 자연스러움을 챙기되, 의미가 여러 가지가 되거나 잘못 해석될 수 있는 경우에는 그대로 두는 방향을 주로 고르고 있어요.

### MVC 디자인 패턴 준수 면에서

전체적으로 Model, Controller, View가 하는 일은 잘 배분하신 것 같아요. 대신 딱 한 곳만 적어보자면, Model에서 View가 가지면 좋을 것 같은 **꽤 구체적인 메시지를** 가지고 있다는 생각이 들었어요. 이 부분은 코멘트로 남겨보았어요.

하나 말씀드려보고 싶은 의견은, Model은 비즈니스 로직에 집중한 후, 결과를 **활용하기 좋은 형태로** 반환하면 어떨까 하는 생각이 들어요. 예를 들면 객체, 배열 등이요. 이러한 데이터들은 당장 출력하기 위해 그대로 가져다 쓸 수는 없지만 가능성을 열어둔 데이터 타입이라고 생각해요.

- 추후 이 Model의 로직을 다른 Model이 가져다 쓸 수도 있고
- Controller가 받을 수 있도 있고
- Controller를 통해 View로 이 데이터 형식이 전달될 수도 있을 거에요.

즉 Model의 결과물은 아주 다양한 곳에서 활용될 수 있어요. 그래서 이에 통용되는 사용하기 좋은 데이터 방식이 좋다고 생각되며, 이후 다른 Controller 및 View 로직이 바뀌더라도 Controller와 View의 관심사를 지니지 않는 Model은 추가 작업이 필요없이 자신이 하는 일에 집중하면 되니 유지보수에 유리할 것이라 생각해요.

`play` 함수의 경우 요구사항 특성상 무조건 스트라이크와 볼의 결과를 반환하거나, 낫싱을 문자열 형태로 반환해야만 해요. 그렇다면 `play` 함수가 하는 일과 다른 아이들이 하는 일들은 어떻게 나눠볼 수 있을까요?

## 3️⃣ 질문

아래의 질문을 하나 해 보고자 해요. 다음 리뷰 요청 시 재홍님의 생각을 말씀해주실 수 있으신가요?

- Q. MVC 패턴에서 굳이 Controller를 사용하여 Model과 View 사이에서 중개하는 방식으로 관리하는 이유가 무엇일까요? Model과 View가 직접 상호작용할 수도 있고, 그러면 굳이 Controller를 중간에 두어 코드량을 늘릴 필요가 없게도 느껴지는데 말이죠.

Commit:
refactor(baseballGameIndex.js,BaseballGameModel.js):통일성과 가독성을 위한 함수명 변경

Commit:
refactor(BaseballGameModel.js,checkInput.js):checkInput.js를 BaseballGameModel.js의 매서드로 등록

isValidInput함수가 model의 기능을 돕는 도매인 함수임을 고려하여 함수 위치 변경

Commit:
refactor(BaseballGameView.js,BaseballGameModel.js,BaseballGameView.js):게임완료 문자열 처리 구역 변경

게임완료 문자열 🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?
이 model에 있는 것이 mvc패턴에 적합하지 않는다 판단하여 control에서 처리하고
같은 이유로 view에서도 같은 문자열 control에서 처리함

Commit:
refactor(BaseballGameView.js)이전 커밋에서 같이 제거 못한 문자열 제거

Commit:
refactor(baseballGameIndex.js,BaseballGameModel.js):play()기능분리 밎 확장된 calculatePlayResult 함수 추가

play()함수를 calculatePlayResult()대체하여 객체를 통한 보편적인 반환을 가진 model함수를 고안하였으나
과제 요건을 고려하여 play함수를 문자열 반환하는 추가기능을 가진 함수로 사용.

Commit:
fix(baseballGameIndex.js,BaseballGameView.js):숫자야구 자리수 변동에 대한 미구현 분제 해결

Commit:
feat(baseballGameIndex.js):사용자가 정답을 맞출 시 더이상 입력을 처리하지 않음

Commit:
refactor(baseballGameIndex.js):BaseballGame클라스 초기호 기능 생성자내부로 통일

Commit:
refactor(baseballGameIndex.js, BaseballGameModel.js, BaseballGameView.js):클라스내 변수,매서드 프로퍼티 캡슐화

Commit:
refactor(baseballGameIndex.js,BaseballGameView.js):view에서 dom 변수로 선언,controller클라스 매게변수 비우기

yoonjaehong26 commented:
김의천 리뷰어님! 올려주신 리뷰 보고 수정해나가면서 많이 배워 나가고 있습니다.
처음 pr을 드릴 때와 비교하여 지금 다시 보니 정말 코드를 보기 좋아진 것 같아 감사드립니다!
***
우선 mvc패턴 적용 이후 가장 크게 느낀점은 관심사 분리를 통한 유지보수의 용이성이 체감된 것입니다.
`'🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?'`
리뷰에서 말씀해 주셨듯이 이 문자열의 처리구역에 대하여 model과 control에서 나누어 다루고 있어 고민이 있었는데
mvc구조를 통해 역할을 명확히 하니 원하는 부분만 검토하고 수정이 가능하여 편리하였습니다.
***
두번째로 주의를 기울인 기능은 실질적인 확장성을 고려한 것 입니다.
김의천 리뷰어님께서 말씀해 주셨듯이 실제로는 자리수의 확장에 대한 코드구성이 미흡하였기에
상상이 아닌 실제로 n자리수 확장을 해보고 코드를 구상하니 확장성에 대해 더 알 수 있었습니다!
***
```
refactor(BaseballGameModel.js,checkInput.js):checkInput.js를 BaseballGameModel.js의 매서드로 등록

isValidInput함수가 model의 기능을 돕는 도매인 함수임을 고려하여 함수 위치 변경
```
위 커밋에서 어떠한 기능에서도 쓰일 수 있는 numToArr함수는 utils폴더에,
숫자 유휴성 검사를 진행하는 checkInput.js는 model의 부가기능이지만 checkInput.js을 통하여 설정값을 변경하는 확장성을 고려하여
별도의 파일로 저장하였습니다.
***
게임 완료 ui관련하여, 숫자를 다 맞출 시 dom을 조작하여 입력창을 없앨까 하였으나 기존 html파일 구조를 유지하는 것이 과제에 맞다고 생각하여
isCompleteGame변수를 통해 추가적인 내부로직에 접근을 막았습니다.
***
말씀해주신 model에서 객체반환을 통한 유동적인 매서드구성에 대하여 고민을 하였고
객체반환을 통한 유동적인 매서드가 정말 좋은 방식이라고 생각하였으나 과제 요건에따라 play()는 문자열을 반환해야 하기에
calculatePlayResult()매서드와 getPlayResultString()의 구성으로 충분하나 play()를 끼워넣은 구조로 보시면 코드를 읽기 편하실 것 같습니다. (이러한 구조에 고민을 많이 하였으나 더 나은 구조를 고민해 보겠습니다)
***
마지막으로 클라스와 캡슐화에 대해 공부해 보아 코드에 적용시켜봤습니다. 
감사합니다! 아래는 4자리 숫자게임 구현 gif입니다.

![화면 기록 2025-09-22 오전 12 13 32](https://github.com/user-attachments/assets/b9d4c476-f4f6-49a5-86e1-d011e76e2981)

wzrabbit commented:
오오, 보아하니 재홍님은 특정 커밋들을 코멘트에 언급하면서 티키타카를 하시는 것을 좋아하시는 것 같아요. 깨알 팁이 있는데 여기 적어보아요. 이 내용은 Discord의 정보 채널에도 적었어요.

## 간혹가다가 리뷰어 또는 동료분께 자신의 커밋을 코멘트에서 강조하고 싶으신 분들을 위한 깨알 팁이에요.
- 각 커밋은 고유한 해시값을 지니는데, 이를 커밋 해시(commit hash)라고 해요.
- 코멘트에 커밋 해시만 적어도, 자동으로 링크로 변환됩니다!

<img width="1163" height="738" alt="image" src="https://github.com/user-attachments/assets/1fa9723b-da23-4bdd-ab1f-f8fdac0502c2" />

<img width="1094" height="204" alt="image" src="https://github.com/user-attachments/assets/eb76ca87-2e4c-4c83-8f2f-a991fe507487" />

<img width="794" height="355" alt="image" src="https://github.com/user-attachments/assets/bcb83b8b-1df2-42aa-b892-79dcd28d5bbd" />

wzrabbit reviewed (approved):
@yoonjaehong26 👋🏻

변경사항 확인하고 적어주신 코멘트에 대해 코멘트 적어보았어요. 오우 코드 전체를 뒤집었다는 말이 어울릴 정도로 많은 부분을 수정해주셨군요! 고생 많으셨어요. 적어주신 코멘트에 대해서 제 의견도 적어보며 리뷰 마칠게요.

> 우선 mvc패턴 적용 이후 가장 크게 느낀점은 관심사 분리를 통한 유지보수의 용이성이 체감된 것입니다.
'🎉축하합니다!!!🎉 정답을 맞추셨습니다. <br> 재시작 하시겠습니까?'
리뷰에서 말씀해 주셨듯이 이 문자열의 처리구역에 대하여 model과 control에서 나누어 다루고 있어 고민이 있었는데
mvc구조를 통해 역할을 명확히 하니 원하는 부분만 검토하고 수정이 가능하여 편리하였습니다.

맞아요. 관심사의 분리는 MVC 패턴의 주요 특징 중 하나에요. 단순히 물건들을 각 바구니에 분류하는 것 그 이상이에요. Model과 View가 각각 자신의 독립적인 기능만을 담당하고, 이를 중계하는 Controller를 거치는 방법을 사용할 때의 장점은 말씀하신 관심사 분리도 있고, 각각의 기능들이 **느슨하게 결합될 수 있는 구조**를 가져가도록 도움을 주어요. 기능들이 느슨하게 결합되었다는 것은 무엇일까요? 바로 기능들이 서로 약하게 연관되어 독립적으로 변경하거나 교체하기 쉬운 상태를 의미해요. 여러 기능들이 복잡하게 얽혀 서로에 크게 의존하고, 하나의 함수가 여러 관심사를 가지면 추후 문제가 발생했을 때 문제의 원인을 찾기 힘들고, 기능을 교체할 때 영향을 받는 기능들이 많은 형태에서는 부작용이 발생하기 쉬워요. **관심사를 나누고 느슨한 구조를 갖춰 유지보수에 용이한 구조를 가져가는 것**은 MVC의 기본적인 목적이라 할 수 있어요. 보다 자세한 이야기는, 제가 이전 리뷰에서 [제 의견](https://github.com/greedy-team/javascript-baseball-precourse/pull/11#pullrequestreview-2353249302) 을 공유해 둔 적이 있어 링크로도 첨부해 드려요.

> 위 커밋에서 어떠한 기능에서도 쓰일 수 있는 numToArr함수는 utils폴더에,
숫자 유휴성 검사를 진행하는 checkInput.js는 model의 부가기능이지만 checkInput.js을 통하여 설정값을 변경하는 확장성을 고려하여
별도의 파일로 저장하였습니다.

잘 구분하신 것 같아요. 이렇게 유틸 함수와 도메인 함수를 구분해 Model을 보조하는 또다른 Model로 만들지, 유틸 함수 디렉토리에 저장할 지도 앞으로 고민해보세요!

> 게임 완료 ui관련하여, 숫자를 다 맞출 시 dom을 조작하여 입력창을 없앨까 하였으나 기존 html파일 구조를 유지하는 것이 과제에 맞다고 생각하여
isCompleteGame변수를 통해 추가적인 내부로직에 접근을 막았습니다.

확인했어요. 게임이 끝난 이후에는 입력창에 무언가를 입력한 후 확인을 눌러도 반영이 되지 않더라고요. 비즈니스 로직도 중요하지만 프론트엔드는 이렇게 사용자와 상호작용하는 UI/UX 부분에도 신경을 쓸 필요가 있어요. 앞으로도 계속 개선해보시길 바래요 💪🏻 

그리고, 기존 HTML 구조를 유지하는 것을 원하신 것 같은데, 꼭 입력창을 없애버려 구조를 깨는 방법 말고도 해결 방법은 있답니다! 아래의 비밀번호 변경 창을 보실래요?

https://github.com/user-attachments/assets/788fb258-d927-4806-9281-acad31c213b2

재홍님과 동일하게 입력창이 작동하지 않도록 하는 방법을 사용했지만, 이쪽은 더 명확하게 **사용자가 이 입력창을 건드릴 수 없음** 이 더 잘 드러나는 예시라고 생각하는데, 재홍님의 생각은 어떠신가요? `<input>`에는 `disabled` 속성을 이용할 수 있는데, `disabled`가 `true`이면 기본적으로 이렇게 입력창이 비활성화되어 건드릴 수 없는 UI가 됩니다.

> 말씀해주신 model에서 객체반환을 통한 유동적인 매서드구성에 대하여 고민을 하였고
객체반환을 통한 유동적인 매서드가 정말 좋은 방식이라고 생각하였으나 과제 요건에따라 play()는 문자열을 반환해야 하기에
calculatePlayResult()매서드와 getPlayResultString()의 구성으로 충분하나 play()를 끼워넣은 구조로 보시면 코드를 읽기 편하실 것 같습니다. (이러한 구조에 고민을 많이 하였으나 더 나은 구조를 고민해 보겠습니다)

좋은 구조 같아요. Model의 반환값을 사용하기 좋은 방법 고민해 보시느라 수고 많이 하셨어요. ?볼 ?스트라이크와 같은 문자열 형태로도 충분히 활용하기 좋은 형태라 생각해요.

이로써 숫자야구 미션을 완료하시게 되었어요. 코멘트를 적는 시점에는 점심 뭐먹지 미션에 첫발을 들이고 계시겠죠? React라는 생소한 라이브러리와 새로운 개념들이 재홍님을 기다리고 있지만, HTML, CSS, JavaScript, 그리고 클린 코드 및 기능 구현을 위해 여러 구조를 만드는 방법은 여전히 계속 사용될 거에요. 숫자야구 미션을 통해 접한 개념들을 자유자재로 활용하시길 바라겠습니다!
